### PR TITLE
Swap out Recoil for Redux + RTK Query

### PR DESCRIPTION
**Pull request by Mrtenz**
This swaps out Recoil for Redux and RTK Query, because:

- Recoil's async selectors don't work well with Gatsby, causing the build to hang.
- Recoil doesn't support reducers out-of-the-box.
- Invalidating Recoil state was a bit difficult.
- React context wasn't really suitable for our use cases (anymore).

### CODE REVIEW

**Comment by Mrtenz** [comment]
@SocketSecurity ignore use-sync-external-store@1.2.0
@SocketSecurity ignore react-redux@8.1.2
@SocketSecurity ignore @reduxjs/toolkit@1.9.6

File: src/features/snaps/Snaps.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import type { FunctionComponent } from 'react';
+
+import { LoadingGrid, SnapCard } from './components';
+import { useSelector } from '../../hooks';
+import { getFilteredSnaps } from '../filter';
+
+export const Snaps: FunctionComponent = () => {
+  const snaps = useSelector(getFilteredSnaps);
+
+  if (snaps.length === 0) {
```
Comment (FrederikBolding) [inline src/features/snaps/Snaps.tsx:11]: Should probably be a loading bool instead or we will show it when searching

File: src/components/InstallSnapButton.tsx [hunk 2]
```diff
@@ -23,39 +24,26 @@ export const InstallSnapButton: FunctionComponent<InstallSnapButtonProps> = ({
...
-  const isInstalled = Boolean(snaps[snapId]);
+  const isInstalled = Boolean(installedSnaps?.[snapId]);
 
   const handleInstall = () => {
-    if (!provider || installing) {
-      return;
-    }
+    installSnap({ snapId, version })
+      .then((result) => {
+        if (hasProperty(result, 'error')) {
```
Comment (FrederikBolding) [inline src/components/InstallSnapButton.tsx]: Shouldn't `installSnap` throw instead?

File: src/components/InstallSnapButton.tsx [hunk 3]
```diff
@@ -23,39 +23,26 @@ export const InstallSnapButton: FunctionComponent<InstallSnapButtonProps> = ({
...
-  const isInstalled = Boolean(snaps[snapId]);
+  const isInstalled = Boolean(installedSnaps?.[snapId]);
 
   const handleInstall = () => {
-    if (!provider || installing) {
-      return;
-    }
+    installSnap({ snapId, version })
+      .then((result) => {
+        if ('error' in result) {
```
Comment (FrederikBolding) [inline src/components/InstallSnapButton.tsx:36]: Shouldn't this be throwing instead?

Comment (Mrtenz) [reply src/components/InstallSnapButton.tsx:36]: The RTK hooks don't throw for some reason.

File: src/features/snaps/Snaps.tsx (hunk 1 above)
Comment (Mrtenz) [reply src/features/snaps/Snaps.tsx:11]: Will do in a follow up PR (empty state).

File: src/components/SnapsProvider.tsx [hunk 4]
```diff
@@ -41,11 +39,10 @@ export const SnapsProvider: FunctionComponent<SnapsProviderProps> = ({
     }
   `);
 
-  const shuffledSnaps = useMemo(() => shuffle(allSnap.nodes), [allSnap.nodes]);
+  useEffect(() => {
+    const shuffledSnaps = shuffle(allSnap.nodes);
+    store.dispatch(setSnaps(shuffledSnaps));
+  }, [store, allSnap.nodes]);
```
[the remark this answers is not in the document]
Comment (Mrtenz) [reply src/components/SnapsProvider.tsx:45]: ESLint disagrees:

> ESLint: React Hook useEffect has a missing dependency: 'store'. Either include it or remove the dependency array.(react-hooks/exhaustive-deps)

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**